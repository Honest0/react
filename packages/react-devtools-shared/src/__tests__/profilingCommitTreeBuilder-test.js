/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import type Store from 'react-devtools-shared/src/devtools/store';

describe('commit tree', () => {
  let React;
  let ReactDOMClient;
  let Scheduler;
  let legacyRender;
  let store: Store;
  let utils;

  beforeEach(() => {
    utils = require('./utils');
    utils.beforeEachProfiling();

    legacyRender = utils.legacyRender;

    store = global.store;
    store.collapseNodesByDefault = false;
    store.recordChangeDescriptions = true;

    React = require('react');
    ReactDOMClient = require('react-dom/client');
    Scheduler = require('scheduler');
  });

  // @reactVersion >= 16.9
  it('should be able to rebuild the store tree for each commit', () => {
    const Parent = ({count}) => {
      Scheduler.unstable_advanceTime(10);
      return new Array(count)
        .fill(true)
        .map((_, index) => <Child key={index} />);
    };
    const Child = React.memo(function Child() {
      Scheduler.unstable_advanceTime(2);
      return null;
    });

    const container = document.createElement('div');

    utils.act(() => store.profilerStore.startProfiling());
    utils.act(() => legacyRender(<Parent count={1} />, container));
    expect(store).toMatchInlineSnapshot(`
      [root]
        ▾ <Parent>
            <Child key="0"> [Memo]
    `);
    utils.act(() => legacyRender(<Parent count={3} />, container));
    expect(store).toMatchInlineSnapshot(`
      [root]
        ▾ <Parent>
            <Child key="0"> [Memo]
            <Child key="1"> [Memo]
            <Child key="2"> [Memo]
    `);
    utils.act(() => legacyRender(<Parent count={2} />, container));
    expect(store).toMatchInlineSnapshot(`
      [root]
        ▾ <Parent>
            <Child key="0"> [Memo]
            <Child key="1"> [Memo]
    `);
    utils.act(() => legacyRender(<Parent count={0} />, container));
    expect(store).toMatchInlineSnapshot(`
      [root]
          <Parent>
    `);
    utils.act(() => store.profilerStore.stopProfiling());

    const rootID = store.roots[0];
    const commitTrees = [];
    for (let commitIndex = 0; commitIndex < 4; commitIndex++) {
      commitTrees.push(
        store.profilerStore.profilingCache.getCommitTree({
          commitIndex,
          rootID,
        }),
      );
    }

    expect(commitTrees[0].nodes.size).toBe(3); // <Root> + <Parent> + <Child>
    expect(commitTrees[1].nodes.size).toBe(5); // <Root> + <Parent> + <Child> x 3
    expect(commitTrees[2].nodes.size).toBe(4); // <Root> + <Parent> + <Child> x 2
    expect(commitTrees[3].nodes.size).toBe(2); // <Root> + <Parent>
  });

  describe('Lazy', () => {
    async function fakeImport(result) {
      return {default: result};
    }

    const LazyInnerComponent = () => null;

    const App = ({renderChildren}) => {
      if (renderChildren) {
        return (
          <React.Suspense fallback="Loading...">
            <LazyComponent />
          </React.Suspense>
        );
      } else {
        return null;
      }
    };

    let LazyComponent;
    beforeEach(() => {
      LazyComponent = React.lazy(() => fakeImport(LazyInnerComponent));
    });

    // @reactVersion >= 16.9
    it('should support Lazy components (legacy render)', async () => {
      const container = document.createElement('div');

      utils.act(() => store.profilerStore.startProfiling());
      utils.act(() => legacyRender(<App renderChildren={true} />, container));
      await Promise.resolve();
      expect(store).toMatchInlineSnapshot(`
        [root]
          ▾ <App>
              <Suspense>
      `);
      utils.act(() => legacyRender(<App renderChildren={true} />, container));
      expect(store).toMatchInlineSnapshot(`
        [root]
          ▾ <App>
            ▾ <Suspense>
                <LazyInnerComponent>
      `);
      utils.act(() => legacyRender(<App renderChildren={false} />, container));
      expect(store).toMatchInlineSnapshot(`
        [root]
            <App>
      `);
      utils.act(() => store.profilerStore.stopProfiling());

      const rootID = store.roots[0];
      const commitTrees = [];
      for (let commitIndex = 0; commitIndex < 3; commitIndex++) {
        commitTrees.push(
          store.profilerStore.profilingCache.getCommitTree({
            commitIndex,
            rootID,
          }),
        );
      }

      expect(commitTrees[0].nodes.size).toBe(3); // <Root> + <App> + <Suspense>
      expect(commitTrees[1].nodes.size).toBe(4); // <Root> + <App> + <Suspense> + <LazyInnerComponent>
      expect(commitTrees[2].nodes.size).toBe(2); // <Root> + <App>
    });

    // @reactVersion >= 18.0
    it('should support Lazy components (createRoot)', async () => {
      const container = document.createElement('div');
      const root = ReactDOMClient.createRoot(container);

      utils.act(() => store.profilerStore.startProfiling());
      utils.act(() => root.render(<App renderChildren={true} />));
      await Promise.resolve();
      expect(store).toMatchInlineSnapshot(`
        [root]
          ▾ <App>
              <Suspense>
      `);
      utils.act(() => root.render(<App renderChildren={true} />));
      expect(store).toMatchInlineSnapshot(`
        [root]
          ▾ <App>
            ▾ <Suspense>
                <LazyInnerComponent>
      `);
      utils.act(() => root.render(<App renderChildren={false} />));
      expect(store).toMatchInlineSnapshot(`
        [root]
            <App>
      `);
      utils.act(() => store.profilerStore.stopProfiling());

      const rootID = store.roots[0];
      const commitTrees = [];
      for (let commitIndex = 0; commitIndex < 3; commitIndex++) {
        commitTrees.push(
          store.profilerStore.profilingCache.getCommitTree({
            commitIndex,
            rootID,
          }),
        );
      }

      expect(commitTrees[0].nodes.size).toBe(3); // <Root> + <App> + <Suspense>
      expect(commitTrees[1].nodes.size).toBe(4); // <Root> + <App> + <Suspense> + <LazyInnerComponent>
      expect(commitTrees[2].nodes.size).toBe(2); // <Root> + <App>
    });

    // @reactVersion >= 16.9
    it('should support Lazy components that are unmounted before resolving (legacy render)', async () => {
      const container = document.createElement('div');

      utils.act(() => store.profilerStore.startProfiling());
      utils.act(() => legacyRender(<App renderChildren={true} />, container));
      expect(store).toMatchInlineSnapshot(`
        [root]
          ▾ <App>
              <Suspense>
      `);
      utils.act(() => legacyRender(<App renderChildren={false} />, container));
      expect(store).toMatchInlineSnapshot(`
        [root]
            <App>
      `);
      utils.act(() => store.profilerStore.stopProfiling());

      const rootID = store.roots[0];
      const commitTrees = [];
      for (let commitIndex = 0; commitIndex < 2; commitIndex++) {
        commitTrees.push(
          store.profilerStore.profilingCache.getCommitTree({
            commitIndex,
            rootID,
          }),
        );
      }

      expect(commitTrees[0].nodes.size).toBe(3); // <Root> + <App> + <Suspense>
      expect(commitTrees[1].nodes.size).toBe(2); // <Root> + <App>
    });

    // @reactVersion >= 18.0
    it('should support Lazy components that are unmounted before resolving (createRoot)', async () => {
      const container = document.createElement('div');
      const root = ReactDOMClient.createRoot(container);

      utils.act(() => store.profilerStore.startProfiling());
      utils.act(() => root.render(<App renderChildren={true} />));
      expect(store).toMatchInlineSnapshot(`
        [root]
          ▾ <App>
              <Suspense>
      `);
      utils.act(() => root.render(<App renderChildren={false} />));
      expect(store).toMatchInlineSnapshot(`
        [root]
            <App>
      `);
      utils.act(() => store.profilerStore.stopProfiling());

      const rootID = store.roots[0];
      const commitTrees = [];
      for (let commitIndex = 0; commitIndex < 2; commitIndex++) {
        commitTrees.push(
          store.profilerStore.profilingCache.getCommitTree({
            commitIndex,
            rootID,
          }),
        );
      }

      expect(commitTrees[0].nodes.size).toBe(3); // <Root> + <App> + <Suspense>
      expect(commitTrees[1].nodes.size).toBe(2); // <Root> + <App>
    });
  });
});
