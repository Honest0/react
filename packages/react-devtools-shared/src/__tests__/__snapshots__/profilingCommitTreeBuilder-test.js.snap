// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit tree Lazy should support Lazy components (createRoot): 0: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Suspense",
      "hocDisplayNames": null,
      "id": 3,
      "key": null,
      "parentID": 2,
      "treeBaseDuration": 0,
      "type": 12,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components (createRoot): 1: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
    3 => Object {
      "children": Array [
        6,
      ],
      "displayName": "Suspense",
      "hocDisplayNames": null,
      "id": 3,
      "key": null,
      "parentID": 2,
      "treeBaseDuration": 0,
      "type": 12,
    },
    6 => Object {
      "children": Array [],
      "displayName": "LazyInnerComponent",
      "hocDisplayNames": null,
      "id": 6,
      "key": null,
      "parentID": 3,
      "treeBaseDuration": 0,
      "type": 5,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components (legacy render): 0: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Suspense",
      "hocDisplayNames": null,
      "id": 3,
      "key": null,
      "parentID": 2,
      "treeBaseDuration": 0,
      "type": 12,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components (legacy render): 1: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
    3 => Object {
      "children": Array [
        6,
      ],
      "displayName": "Suspense",
      "hocDisplayNames": null,
      "id": 3,
      "key": null,
      "parentID": 2,
      "treeBaseDuration": 0,
      "type": 12,
    },
    6 => Object {
      "children": Array [],
      "displayName": "LazyInnerComponent",
      "hocDisplayNames": null,
      "id": 6,
      "key": null,
      "parentID": 3,
      "treeBaseDuration": 0,
      "type": 5,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components that are unmounted before resolving (createRoot): 0: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Suspense",
      "hocDisplayNames": null,
      "id": 3,
      "key": null,
      "parentID": 2,
      "treeBaseDuration": 0,
      "type": 12,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components that are unmounted before resolving (createRoot): 1: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components that are unmounted before resolving (legacy render): 0: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Suspense",
      "hocDisplayNames": null,
      "id": 3,
      "key": null,
      "parentID": 2,
      "treeBaseDuration": 0,
      "type": 12,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree Lazy should support Lazy components that are unmounted before resolving (legacy render): 1: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 0,
      "type": 11,
    },
    2 => Object {
      "children": Array [],
      "displayName": "App",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 0,
      "type": 5,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree should be able to rebuild the store tree for each commit: 0: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 12,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
      ],
      "displayName": "Parent",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 12,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Child",
      "hocDisplayNames": null,
      "id": 3,
      "key": "0",
      "parentID": 2,
      "treeBaseDuration": 2,
      "type": 8,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree should be able to rebuild the store tree for each commit: 1: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 16,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
        4,
        5,
      ],
      "displayName": "Parent",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 16,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Child",
      "hocDisplayNames": null,
      "id": 3,
      "key": "0",
      "parentID": 2,
      "treeBaseDuration": 2,
      "type": 8,
    },
    4 => Object {
      "children": Array [],
      "displayName": "Child",
      "hocDisplayNames": null,
      "id": 4,
      "key": "1",
      "parentID": 2,
      "treeBaseDuration": 2,
      "type": 8,
    },
    5 => Object {
      "children": Array [],
      "displayName": "Child",
      "hocDisplayNames": null,
      "id": 5,
      "key": "2",
      "parentID": 2,
      "treeBaseDuration": 2,
      "type": 8,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree should be able to rebuild the store tree for each commit: 2: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 14,
      "type": 11,
    },
    2 => Object {
      "children": Array [
        3,
        4,
      ],
      "displayName": "Parent",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 14,
      "type": 5,
    },
    3 => Object {
      "children": Array [],
      "displayName": "Child",
      "hocDisplayNames": null,
      "id": 3,
      "key": "0",
      "parentID": 2,
      "treeBaseDuration": 2,
      "type": 8,
    },
    4 => Object {
      "children": Array [],
      "displayName": "Child",
      "hocDisplayNames": null,
      "id": 4,
      "key": "1",
      "parentID": 2,
      "treeBaseDuration": 2,
      "type": 8,
    },
  },
  "rootID": 1,
}
`;

exports[`commit tree should be able to rebuild the store tree for each commit: 3: CommitTree 1`] = `
Object {
  "nodes": Map {
    1 => Object {
      "children": Array [
        2,
      ],
      "displayName": null,
      "hocDisplayNames": null,
      "id": 1,
      "key": null,
      "parentID": 0,
      "treeBaseDuration": 10,
      "type": 11,
    },
    2 => Object {
      "children": Array [],
      "displayName": "Parent",
      "hocDisplayNames": null,
      "id": 2,
      "key": null,
      "parentID": 1,
      "treeBaseDuration": 10,
      "type": 5,
    },
  },
  "rootID": 1,
}
`;
